torch/optim/zeroptimizer.py

```diff
@@ -0,0 +1,524 @@
+                    self._per_device_params[device][self.param_to_rank[param]] += [param]
+
+            # Sort param_lists by size
+            for k in self._per_device_params.keys():
+                for r in self._per_device_params[k]:
+                    r.sort(key=lambda x: x.numel())
+
+        return self._per_device_params
+
+    @property
+    def param_to_rank(self) -> Dict[torch.Tensor, int]:
+        """param to data parallel rank"""
+        if len(self._param_rank) == 0:
+            for rank, param_groups in enumerate(self.partition_parameters()):
+                for param_group in param_groups:
+                    for param in param_group["params"]:
+                        self._param_rank[param] = rank
+        return self._param_rank
+
+    def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) -> Optional[float]:
+        """Performs a single optimization step (parameter update).
+
+        Arguments:
+            closure (callable): A closure that reevaluates the model and
+                returns the loss. Optional for most optimizers.
+
+        .. note: Any extra parameter is passed to the base optimizer as-is"""
+
+        # Sync oss param_groups attributes in case they've been updated by a scheduler.
+        self._sync_param_groups()
+
+        # Run the optimizer step on this shard only:
+        if closure is not None:
+            loss = self.optim.step(closure=closure, **kwargs)  # type: ignore
+        else:
+            loss = self.optim.step(**kwargs)
+
+        # Free the other grads
+        # ideally this would not be needed, handled in the reduce step by a zero-matching DDP engin
+        for rank, partition in enumerate(self.partition_parameters()):
+            if rank != self.rank:
+                for p in partition:
+                    for t in p["params"]:
+                        t.grad = None
+
+        # Sync all the updated shards in between the ranks
+        with torch.no_grad():
+            for (
+                device,
+                device_params,
+            ) in self.per_device_params.items():  # all the params on this device (inc all ranks)
+                self._broadcast_params(self._broadcast_buffers[device], device_params)
+
+        # Sync hypothethical new results from the wrapped optimizer to the exposed param_groups
+        self._sync_param_groups(local_to_global=True)
+
+        return loss
+
+    def local_state_dict(self) -> dict:
+        """Gets this rank's state_dict.
+
+        Returns:
+            The state of the optimizer as a :class:`dict`.
+            It contains two entries:
+
+            * state - a dict holding current optimization state. Its content
+                differs between optimizer classes.
+            * param_groups - a dict containing all parameter groups
+        """
+        return self.optim.state_dict()
+
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
```

Comment:
to start discussions, I don't like that part, I see a couple of issues:
- the API is not the same as normal pytorch optimizers
- this is very slow (broadcast state shards everywhere)
- this adds a fairly big code footprint for a single usage (checkpoints)

I think that relying on rpc/rref would be nicer, but it would require both c10 and rpc to be initialized, very open to other ideas

Reply:
Hey @jeffra @samyam, how does ZeRO solve the state saving problem?

Reply:
@mrshenli, during model checkpointing, we save individual states of each data parallel process involved in ZeRO into a separate file. During the checkpoint loading, we do the consolidation and do the partitioning based on the number of data parallel process. 

In DeepSpeed ZeRO, since we handle the model checkpoint and restore internally, we do not expose any state consolidation APIs to the user. Regardless of whether the user is using ZeRO or not, we have a common, 'save_checkpoint' and 'load_checkpoint' APIs that handles the consolidation in the background if needed. 

I feel like what you guys have is a good solution. Alternatively, you could remove the consolidate API, and just use state_dict() to always return the consolidated states. I feel that since the model parameters themselves are not partitioned for ZeRO Stage 1 and Stage 2, it also makes sense to return a consolidated state dictionary for the optimizer states.  

Another alternative could be to use state_dict(self, consolidation=False), which would make state_dict compatible with all other optimizers, but for ZeRO, you have the option of consolidation.